File: sdk/r/R/stats.R
#' Standalone Wrapper Functions for Statistical Tests
#'
#' These functions provide a functional (non-OOP) interface to the StickForStats
#' API. Each function accepts an optional `client` argument; if omitted, a
#' default client is constructed from saved configuration (see
#' [sfs_configure()]).
#'
#' @name stats-wrappers
NULL

#' Perform a t-test via StickForStats API
#'
#' Independent or paired-samples t-test with Guardian assumption validation.
#'
#' @param data Named list of numeric vectors or a data.frame. For two-sample
#'   tests, provide two groups (e.g. `list(control = c(...), treatment = c(...))`).
#' @param groups Character vector. Explicit group names to compare.
#' @param paired Logical. Whether to run a paired-samples test.
#' @param alpha Numeric. Significance level (default 0.05).
#' @param alternative Character. `"two-sided"`, `"less"`, or `"greater"`.
#' @param ... Additional parameters passed to the API.
#' @param client A `StickForStats` R6 object. If NULL, uses [sfs_default_client()].
#'
#' @return A list with class `"sfs_result"` containing t-test results including
#'   test statistic, p-value, confidence interval, effect size, and Guardian
#'   validation report.
#'
#' @examples
#' \dontrun{
#' result <- sfs_ttest(
#'   data = list(control = c(5.1, 4.8, 5.3), treatment = c(7.2, 6.9, 7.5)),
#'   alpha = 0.05
#' )
#' print(result)
#' }
#'
#' @export
sfs_ttest <- function(data, groups = NULL, paired = FALSE,
                      alpha = 0.05, alternative = "two-sided",
                      ..., client = NULL) {
  client <- client %||% sfs_default_client()
  client$ttest(data = data, groups = groups, paired = paired,
               alpha = alpha, alternative = alternative, ...)
}


#' Perform an ANOVA via StickForStats API
#'
#' One-way, repeated-measures, or MANOVA with post-hoc tests and Guardian
#' validation.
#'
#' @param data Named list of numeric vectors or a data.frame.
#' @param groups Character vector. Subset of groups to compare.
#' @param alpha Numeric. Significance level.
#' @param anova_type Character. `"one-way"`, `"repeated-measures"`, or
#'   `"manova"`.
#' @param post_hoc Character. Post-hoc method: `"tukey"`, `"bonferroni"`,
#'   `"scheffe"`.
#' @param ... Additional parameters passed to the API.
#' @param client A `StickForStats` R6 object. If NULL, uses [sfs_default_client()].
#'
#' @return A list with class `"sfs_result"` containing ANOVA results.
#'
#' @examples
#' \dontrun{
#' result <- sfs_anova(
#'   data = list(A = c(1,2,3), B = c(4,5,6), C = c(7,8,9)),
#'   post_hoc = "tukey"
#' )
#' }
#'
#' @export
sfs_anova <- function(data, groups = NULL, alpha = 0.05,
                      anova_type = "one-way", post_hoc = "tukey",
                      ..., client = NULL) {
  client <- client %||% sfs_default_client()
  client$anova(data = data, groups = groups, alpha = alpha,
               anova_type = anova_type, post_hoc = post_hoc, ...)
}


#' Compute a correlation coefficient via StickForStats API
#'
#' Pearson, Spearman, or Kendall correlation with significance test.
#'
#' @param x Numeric vector.
#' @param y Numeric vector.
#' @param method Character. `"pearson"`, `"spearman"`, or `"kendall"`.
#' @param alpha Numeric. Significance level for confidence interval.
#' @param ... Additional parameters passed to the API.
#' @param client A `StickForStats` R6 object. If NULL, uses [sfs_default_client()].
#'
#' @return A list with class `"sfs_result"` containing correlation results.
#'
#' @examples
#' \dontrun{
#' result <- sfs_correlation(
#'   x = c(1, 2, 3, 4, 5),
#'   y = c(2, 4, 5, 4, 5),
#'   method = "pearson"
#' )
#' }
#'
#' @export
sfs_correlation <- function(x, y, method = "pearson", alpha = 0.05,
                            ..., client = NULL) {
  client <- client %||% sfs_default_client()
  client$correlation(x = x, y = y, method = method, alpha = alpha, ...)
}


#' Fit a regression model via StickForStats API
#'
#' Supports linear, multiple, polynomial, logistic, ridge, and lasso
#' regression with 50-digit precision.
#'
#' @param data A data.frame or named list of numeric vectors.
#' @param dependent Character. Name of the dependent variable.
#' @param predictors Character vector. Names of predictor variables.
#' @param type Character. Regression type: `"linear"`, `"multiple"`,
#'   `"polynomial"`, `"logistic"`, `"ridge"`, or `"lasso"`.
#' @param alpha Numeric. Significance level.
#' @param ... Additional parameters passed to the API.
#' @param client A `StickForStats` R6 object. If NULL, uses [sfs_default_client()].
#'
#' @return A list with class `"sfs_result"` containing regression results.
#'
#' @examples
#' \dontrun{
#' df <- data.frame(x = 1:10, y = 2 * (1:10) + rnorm(10))
#' result <- sfs_regression(df, dependent = "y", predictors = "x")
#' }
#'
#' @export
sfs_regression <- function(data, dependent, predictors,
                           type = "linear", alpha = 0.05,
                           ..., client = NULL) {
  client <- client %||% sfs_default_client()
  client$regression(data = data, dependent = dependent,
                    predictors = predictors, type = type,
                    alpha = alpha, ...)
}


#' Compute descriptive statistics via StickForStats API
#'
#' Mean, median, standard deviation, skewness, kurtosis, and more.
#'
#' @param data Numeric vector, named list, or data.frame.
#' @param ... Additional parameters passed to the API.
#' @param client A `StickForStats` R6 object. If NULL, uses [sfs_default_client()].
#'
#' @return A list with class `"sfs_result"` containing descriptive statistics.
#'
#' @examples
#' \dontrun{
#' result <- sfs_descriptive(c(10, 20, 30, 40, 50))
#' }
#'
#' @export
sfs_descriptive <- function(data, ..., client = NULL) {
  client <- client %||% sfs_default_client()
  client$descriptive(data = data, ...)
}

File: sdk/python/src/stickforstats/stats.py
"""
Statistics module — wraps ``/api/v1/stats/*`` endpoints.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from stickforstats.models import (
    ANOVAResult,
    CorrelationResult,
    DescriptiveResult,
    RegressionResult,
    TTestResult,
)

if TYPE_CHECKING:
    from stickforstats.client import StickForStats


class StatsModule:
    """
    Parametric statistical tests with Guardian protection.

    Accessed via ``client.stats``.
    """

    def __init__(self, client: StickForStats) -> None:
        self._client = client

    # ------------------------------------------------------------------
    # t-test
    # ------------------------------------------------------------------

    def ttest(
        self,
        data: dict[str, list[float]],
        *,
        groups: list[str] | None = None,
        paired: bool = False,
        alpha: float = 0.05,
        alternative: str = "two-sided",
        **kwargs: Any,
    ) -> TTestResult:
        """
        Perform an independent or paired-samples t-test.

        Parameters
        ----------
        data : dict
            Mapping of group names to numeric lists, e.g.
            ``{"control": [1.2, 3.4], "treatment": [5.6, 7.8]}``.
        groups : list of str, optional
            Explicit group names to compare.  Defaults to the first two keys
            of *data*.
        paired : bool
            Whether to run a paired-samples t-test.
        alpha : float
            Significance level.
        alternative : str
            ``"two-sided"``, ``"less"``, or ``"greater"``.

        Returns
        -------
        TTestResult
        """
        payload: dict[str, Any] = {
            "data": data,
            "paired": paired,
            "alpha": alpha,
            "alternative": alternative,
            **kwargs,
        }
        if groups is not None:
            payload["groups"] = groups

        resp = self._client.post("stats/ttest/", json=payload)
        return TTestResult.model_validate(resp)

    # ------------------------------------------------------------------
    # ANOVA
    # ------------------------------------------------------------------

    def anova(
        self,
        data: dict[str, list[float]],
        *,
        groups: list[str] | None = None,
        alpha: float = 0.05,
        anova_type: str = "one-way",
        post_hoc: str = "tukey",
        **kwargs: Any,
    ) -> ANOVAResult:
        """
        Perform a one-way (or repeated-measures / MANOVA) ANOVA.

        Parameters
        ----------
        data : dict
            Mapping of group names to numeric lists.
        groups : list of str, optional
            Subset of groups to compare.
        alpha : float
            Significance level.
        anova_type : str
            ``"one-way"``, ``"repeated-measures"``, or ``"manova"``.
        post_hoc : str
            Post-hoc method (``"tukey"``, ``"bonferroni"``, ``"scheffe"``).

        Returns
        -------
        ANOVAResult
        """
        payload: dict[str, Any] = {
            "data": data,
            "alpha": alpha,
            "anova_type": anova_type,
            "post_hoc": post_hoc,
            **kwargs,
        }
        if groups is not None:
            payload["groups"] = groups

        resp = self._client.post("stats/anova/", json=payload)
        return ANOVAResult.model_validate(resp)

    # ------------------------------------------------------------------
    # ANCOVA
    # ------------------------------------------------------------------

    def ancova(
        self,
        data: dict[str, Any],
        *,
        dependent: str,
        factor: str,
        covariates: list[str] | None = None,
        alpha: float = 0.05,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """
        Perform an Analysis of Covariance.

        Parameters
        ----------
        data : dict
            Column-oriented dataset.
        dependent : str
            Name of the dependent variable.
        factor : str
            Name of the grouping factor.
        covariates : list of str, optional
            Covariate column names.
        alpha : float
            Significance level.

        Returns
        -------
        dict
            Raw API response.
        """
        payload: dict[str, Any] = {
            "data": data,
            "dependent": dependent,
            "factor": factor,
            "alpha": alpha,
            **kwargs,
        }
        if covariates is not None:
            payload["covariates"] = covariates

        return self._client.post("stats/ancova/", json=payload)

    # ------------------------------------------------------------------
    # Correlation
    # ------------------------------------------------------------------

    def correlation(
        self,
        x: list[float],
        y: list[float],
        *,
        method: str = "pearson",
        alpha: float = 0.05,
        **kwargs: Any,
    ) -> CorrelationResult:
        """
        Compute a correlation coefficient with significance test.

        Parameters
        ----------
        x, y : list of float
            Paired numeric samples.
        method : str
            ``"pearson"``, ``"spearman"``, or ``"kendall"``.
        alpha : float
            Significance level for confidence interval.

        Returns
        -------
        CorrelationResult
        """
        payload: dict[str, Any] = {
            "x": x,
            "y": y,
            "method": method,
            "alpha": alpha,
            **kwargs,
        }
        resp = self._client.post("stats/correlation/", json=payload)
        return CorrelationResult.model_validate(resp)

    # ------------------------------------------------------------------
    # Regression
    # ------------------------------------------------------------------

    def regression(
        self,
        data: dict[str, list[float]],
        *,
        dependent: str,
        predictors: list[str],
        regression_type: str = "linear",
        alpha: float = 0.05,
        **kwargs: Any,
    ) -> RegressionResult:
        """
        Fit a regression model.

        Parameters
        ----------
        data : dict
            Column-oriented dataset.
        dependent : str
            Name of the dependent variable column.
        predictors : list of str
            Names of predictor columns.
        regression_type : str
            ``"linear"``, ``"multiple"``, ``"polynomial"``, ``"logistic"``,
            ``"ridge"``, or ``"lasso"``.
        alpha : float
            Significance level.

        Returns
        -------
        RegressionResult
        """
        payload: dict[str, Any] = {
            "data": data,
            "dependent": dependent,
            "predictors": predictors,
            "regression_type": regression_type,
            "alpha": alpha,
            **kwargs,
        }
        resp = self._client.post("stats/regression/", json=payload)
        return RegressionResult.model_validate(resp)

    # ------------------------------------------------------------------
    # Descriptive statistics
    # ------------------------------------------------------------------

    def descriptive(
        self,
        data: list[float] | dict[str, list[float]],
        **kwargs: Any,
    ) -> DescriptiveResult:
        """
        Compute descriptive statistics (mean, median, SD, etc.).

        Parameters
        ----------
        data : list of float or dict
            A flat list of numbers, or a column-oriented dict.

        Returns
        -------
        DescriptiveResult
        """
        payload: dict[str, Any] = {"data": data, **kwargs}
        resp = self._client.post("stats/descriptive/", json=payload)
        return DescriptiveResult.model_validate(resp)
